### PR TITLE
Extract common functions for compilation

### PR DESCRIPTION
**Pull request by fzyzcjy**
## Motivation

this should be safe b/c I do not change any computation, but only extract code to functions (the next PR will change computation separately)

## Checklist

- [ ] Format your code according to the [Code Formatting with Pre-Commit](https://docs.sglang.ai/references/contribution_guide.html#code-formatting-with-pre-commit).
- [ ] Add unit tests as outlined in the [Running Unit Tests](https://docs.sglang.ai/references/contribution_guide.html#running-unit-tests-adding-to-ci).
- [ ] Update documentation / docstrings / example tutorials as needed, according to [Writing Documentation](https://docs.sglang.ai/references/contribution_guide.html#writing-documentation-running-docs-ci).
- [ ] Provide throughput / latency benchmark results and accuracy evaluation results as needed, according to [Benchmark and Profiling](https://docs.sglang.ai/references/benchmark_and_profiling.html) and [Accuracy Results](https://docs.sglang.ai/references/accuracy_evaluation.html).
- [ ] For reviewers: If you haven't made any contributions to this PR and are only assisting with merging the main branch, please remove yourself as a co-author when merging the PR.
- [ ] Please feel free to join our Slack channel at https://sgl-fru7574.slack.com/archives/C09784E3EN6 to discuss your PR.

### CODE REVIEW

File: specforge/core/eagle3.py [hunk 1]
```diff
@@ -447,3 +435,33 @@ def forward(
...
+    plogp = target_p * out_logp
+    loss = -torch.sum(position_mask * plogp, 2).mean()
+    return loss
+
+
+@torch.compile(dynamic=None)
+def _compute_metric_acc(logits, target_p, position_mask, loss_mask):
+    return (
+        (logits.argmax(-1) == target_p.argmax(-1)) * position_mask.squeeze(-1)
+    ).sum().item() / (loss_mask.sum().item() + 1e-6)
```
[the remark this answers is not in the document]
Comment (yubofredwang) [reply specforge/core/eagle3.py:467]: I think this is a valid comment. Is it intentional to remove the float conversion?

File: specforge/core/eagle3.py [hunk 2]
```diff
@@ -447,3 +435,33 @@ def forward(
                     ).min
                 # Flex attention mask shirnking is handled inside attention module
         return plosses, vlosses, acces
+
+
+@torch.compile(dynamic=None)
```
Comment (yubofredwang) [inline specforge/core/eagle3.py:440]: sequence length can vary from batch to batch, wondering if we can turn dynamic=True for this?

Did you notice compile overhead when testing with dynamic=None?

Comment (fzyzcjy) [reply specforge/core/eagle3.py:440]: from my experience it is good and no recompile, maybe the dynamic=none is enough (iirc it detects what dim change and mark that specific dim as dynamic and thus only 1 recompile at start)

File: specforge/core/eagle3.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (fzyzcjy) [reply specforge/core/eagle3.py:467]: Firstly, yes that to-float is fused into compute_loss and the _compute_metric_acc does have a change. Secondly, for here (the acc part), it seems `logits.argmax()` will be the same for bf16 or upcasted-fp32, thus it seems there is no worries.

**Comment by yubofredwang** [comment]
Running the training e2e to test out the PR.

**Comment by fzyzcjy** [comment]
no worries I have run one (that includes this code change and a future pr code change etc): https://github.com/sgl-project/SpecForge/pull/157
